### ISSUE TITLE
Hopperhock filter (item frame) does not function at far coordinates.

### ISSUE THREAD

**Issue by OJwithC** [Vazkii/Botania#1251]
Botania v208

Somewhere just before x530,000 z530,000 (same for negative values), item frame filters, for the hopperhock at least, stop working. If you put an item frame on an adjacent inventory to filter it, the hopperhock picks up everything around it regardless of the flower's setting or the items in the frame. I found this out because the spread command for Garden of Glass put me past those coordinates.

**Comment by Vazkii** [comment]
Ok, this is really weird.

**Comment by Ellpeck** [comment]
Is the position of the Item Frame direction-dependant or does it happen on any side of the chest? :O

**Comment by Vazkii** [comment]
I can reproduce it at any side at (55002, 86, 550002).

**Comment by Kolatra** [comment]
I'll try as well in my dev environment. This is definitely odd...

**Comment by Speiger** [comment]
I would like to help but it would be nice where i can find the hopperhock flower class (logic)

**Comment by Ellpeck** [comment]
https://github.com/Vazkii/Botania/blob/master/src/main/java/vazkii/botania/common/block/subtile/functional/SubTileHopperhock.java

**Comment by Speiger** [comment]
hmmm no clue... Nothing seem to fit there...

**Comment by Ellpeck** [comment]
I'm currently looking at getEntitiesWithinAABB. I don't really get World's MAX_ENTITY_RADIUS.. seems pretty weird to me.

**Comment by lumien231** [comment]
At these high coordinates the chunk coordinates extend the range of a short (32.767). Maybe that has something to do with it?

**Comment by 0xc6** [comment]
I am almost certain that this is due to the way in which Minecraft handles entity locations. Due to floating-point inaccuracies, the further that one travels from (0, 0), the less accurate the location of an entity is. Because the logical location of the Item Frame isn't necessarily where it is rendered, Botania doesn't actually see it to be within range of the inventory.

TL;DR: Entities act weirdly at high coordinates.

Well, that's my theory, at least.

**Comment by TTFTCUTS** [comment]
As 0xc6 said, this is due to floating point inaccuracy. Minecraft starts going weird around 100k, and gets worse the further out you go. Players are never meant to get anywhere near that far out under normal circumstances.

See here: [MC wiki's farlands article, effects section](http://minecraft.gamepedia.com/Farlands#Effects) for more information on the effects of being too far out.

**Comment by Ellpeck** [comment]
That's pretty old facts from before 1.8 though, and entities don't really start lagging and glitching around anymore as before. So I thought the floating point issue got fixed (at least somewhat) too :O

**Comment by ljfa-ag** [comment]
It has probably something to do with floating point inaccuracies. <tt>EntityHanging.setDirection</tt> mostly uses floats for bounding box calculations, which are too inaccurate for some of the calculations after 524288.

**Comment by TTFTCUTS** [comment]
Some of the issues were, but I'm willing to bet that by that far out it's not finding the frames due to rounding.

**Comment by Ellpeck** [comment]
Yea, that's probably what it is then.. might not be able to fix it in a good way then.

**Comment by ljfa-ag** [comment]
One resort might be to ditch the bounding box calculations done in <tt>EntityHanging.setDirection</tt> and redo them with doubles after the item frame has spawned. This should lead to a more accurate bounding box.

**Comment by 0xc6** [comment]
@Ellpeck This is still an issue as of 1.7.10/1.8.7, although the rendering of item frames appears to be fixed as of 1.7. References: [MC-3718](https://bugs.mojang.com/browse/MC-3718) and [MC-3493](https://bugs.mojang.com/browse/MC-3493). Essentially, even if they look as if they are in the correct place, Minecraft is just applying some rendering hacks to make them look as if they are located correctly.

**Comment by trabpukcip** [comment]
I was going to say something 4 hours ago just after Vazkii for commented, I would have thought the floating point errors would have been fixed by now?
